inst/include/dust/interface.hpp

```diff
@@ -402,34 +435,31 @@ cpp11::sexp dust_filter(SEXP ptr, bool save_trajectories,
     cpp11::stop("Data has not been set for this object");
   }
 
-  dust::filter::filter_state<real_t> filter_state;
-
   std::vector<size_t> step_snapshot =
-    dust::interface::check_step_snapshot(r_step_snapshot, obj->data());
-  cpp11::writable::doubles
-    log_likelihood(dust::filter::filter(obj, filter_state,
-                                        save_trajectories, step_snapshot));
+      dust::interface::check_step_snapshot(r_step_snapshot, obj->data());
 
+  cpp11::writable::doubles log_likelihood;
   cpp11::sexp r_trajectories, r_snapshots;
-
-  if (save_trajectories) {
-    const auto& trajectories = filter_state.trajectories;
-    cpp11::writable::doubles trajectories_data(trajectories.size());
-    trajectories.history(REAL(trajectories_data));
-    trajectories_data.attr("dim") =
-      dust::interface::state_array_dim(obj->n_state(), obj->shape(),
-                                       obj->n_data() + 1);
-    r_trajectories = trajectories_data;
-  }
-
-  if (r_step_snapshot != R_NilValue) {
-    const auto& snapshots = filter_state.snapshots;
-    cpp11::writable::doubles snapshots_data(snapshots.size());
-    snapshots.history(REAL(snapshots_data));
-    snapshots_data.attr("dim") =
-      dust::interface::state_array_dim(obj->n_state_full(), obj->shape(),
-                                       step_snapshot.size());
-    r_snapshots = snapshots_data;
+  if (device) {
```

Comment:
this might be clearer if the identical inner parts are written as a free function templated over the filter state